// es-dev-server/dist/utils/polyfills.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getPolyfills = getPolyfills;
exports.legacyPolyfills = exports.modernPolyfills = exports.esmPolyfills = void 0;

var _constants = require("../constants.js");

/** @type {import('@open-wc/building-utils/index-html/create-index-html').PolyfillsConfig} */
const esmPolyfills = {
  esModuleShims: true
};
/** @type {import('@open-wc/building-utils/index-html/create-index-html').PolyfillsConfig} */

exports.esmPolyfills = esmPolyfills;
const modernPolyfills = { ...esmPolyfills,
  webcomponents: true
};
/** @type {import('@open-wc/building-utils/index-html/create-index-html').PolyfillsConfig} */

exports.modernPolyfills = modernPolyfills;
const legacyPolyfills = { ...modernPolyfills,
  coreJs: true,
  regeneratorRuntime: true,
  fetch: true,
  systemJsExtended: true
};
/**
 * @param {string} compatibilityMode
 */

exports.legacyPolyfills = legacyPolyfills;

function getPolyfills(compatibilityMode) {
  switch (compatibilityMode) {
    case _constants.compatibilityModes.ESM:
      return esmPolyfills;

    case _constants.compatibilityModes.MODERN:
      return modernPolyfills;

    case _constants.compatibilityModes.ALL:
      return legacyPolyfills;

    default:
      return {};
  }
}
